lib/array.js

```diff
@@ -38,6 +38,63 @@ const map = (
   }
 };
 
+const DEFAULT_OPTIONS = { min: 5, percent: 0.7 };
+
+const asyncMap = (
+  // Non-blocking map
+  items, // array, incoming
+  fn, // function, (value, index) => {}
+  options, // object (optional), { min, percent }
+  done // function (optional), on done callback function(result)
+) => {
+  if (typeof options === 'function') {
+    done = options;
+    options = DEFAULT_OPTIONS;
+  }
+
+  if (!items.length) {
+    if (done) done([]);
+    return;
+  }
+
+  const min = options.min || DEFAULT_OPTIONS.min;
+  const percent = options.percent || DEFAULT_OPTIONS.percent;
+
+  let begin;
+  let sum = 0;
+  let count = 0;
+
+  const relation = percent / (1 - percent);
+  const newArray = new Array(items.length);
+
+  const countNumber = () => {
+    const loopTime = Date.now() - begin;
+    const itemTime = sum / count;
+    const necessaryNumber = relation * loopTime / itemTime;
```

Comment:
I'm not sure I follow, how will this achieve a percentage of total event loop time?
i.e.
we pass percentage 0.2 -> 
  relation = 0.25 ->
  let loopTime be 1000 and itemTime 10 (min is 10 and sum is 100) -> 
  necessaryNumber = 0.25 * 1000 / 10 = 25

which is if we execute will be 25 * 10 = 250 millis
which is 250/1000 0.25 (25%) while we have passed 0.2(20%)
(if we continue => sum 350, looptime 1150 (stable) -> necessaryNumber = 28.75)

are you sure that's correct or am I misunderstanding something?

Reply:
loopTime + mapTime = allTime
250 / (1000 + 250) = 0.2

Reply:
Oh, haven't noticed the reassignment on the line 88, I'd prefer to have a separate variable for for-loop-time though.